cluster-autoscaler/simulator/csi/provider/csinode_test.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	storagev1 "k8s.io/api/storage/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes/fake"
+	v1storagelister "k8s.io/client-go/listers/storage/v1"
+)
+
+// testCSINodeLister is a fake CSINode lister for testing.
+type testCSINodeLister struct {
+	csiNodes []*storagev1.CSINode
+}
+
+func (l *testCSINodeLister) List(selector labels.Selector) ([]*storagev1.CSINode, error) {
+	return l.csiNodes, nil
+}
+
+func (l *testCSINodeLister) Get(name string) (*storagev1.CSINode, error) {
+	for _, csiNode := range l.csiNodes {
+		if csiNode.Name == name {
+			return csiNode, nil
+		}
+	}
+	return nil, apierrors.NewNotFound(storagev1.Resource("csinode"), name)
+}
+
+func newTestCSINodeLister(csiNodes []*storagev1.CSINode) v1storagelister.CSINodeLister {
+	return &testCSINodeLister{csiNodes: csiNodes}
+}
+
+func TestNewCSINodeProvider(t *testing.T) {
+	csiNode1 := &storagev1.CSINode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node-1",
+		},
+		Spec: storagev1.CSINodeSpec{
+			Drivers: []storagev1.CSINodeDriver{
+				{
+					Name:         "driver1",
+					NodeID:       "node-id-1",
+					TopologyKeys: []string{"topology-key-1"},
+				},
+			},
+		},
+	}
+
+	csiNode2 := &storagev1.CSINode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node-2",
+		},
+		Spec: storagev1.CSINodeSpec{
+			Drivers: []storagev1.CSINodeDriver{
+				{
+					Name:         "driver2",
+					NodeID:       "node-id-2",
+					TopologyKeys: []string{"topology-key-2"},
+				},
+			},
+		},
+	}
+
+	csiNodes := []*storagev1.CSINode{csiNode1, csiNode2}
+	lister := newTestCSINodeLister(csiNodes)
+
+	provider := NewCSINodeProvider(lister)
+	require.NotNil(t, provider)
+
+	snapshot, err := provider.Snapshot()
+	require.NoError(t, err)
+	require.NotNil(t, snapshot)
+
+	// Verify that snapshot contains all CSINodes
+	snapshotCSINodes := snapshot.CSINodes()
+	require.NotNil(t, snapshotCSINodes)
```

Comment:
Why not verify the actual result instead of just checking for nil?

Reply:
I have moved this to table drivern tests. But the code *does* test the result too later in the file.